src/nvim/api/autocmd.c

```diff
@@ -0,0 +1,271 @@
+#include <stdbool.h>
+
+#include "lauxlib.h"
+#include "nvim/api/autocmd.h"
+#include "nvim/api/private/defs.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/fileio.h"
+#include "nvim/lua/executor.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "api/autocmd.c.generated.h"
+#endif
+
+#define FOREACH_KEY_VALUE(d, k, v, code) \
+  for (size_t i = 0; i < d.size; i++) { \
+    String k = d.items[i].key; \
+    Object *v = &d.items[i].value; \
+    code; \
+  }
+
+#define FOREACH_ITEM(a, item, code) \
+  for (size_t i = 0; i < a.size; i++) { \
+    Object item = a.items[i]; \
+    code; \
+  }
+
+#define UNPACK_EVENTS(event_array, v, goto_name) \
+  if (v->type == kObjectTypeString) { \
+    ADD(event_array, copy_object(*v)); \
+  } else if (v->type == kObjectTypeArray) { \
+    for (size_t j = 0; j < event_array.size; j++) { \
+      Object item = event_array.items[j]; \
+      if (item.type != kObjectTypeString) { \
+        api_set_error( \
+            err, \
+            kErrorTypeValidation, \
+            "All entries in 'event' must be strings"); \
+        goto goto_name; \
+      } \
+    } \
+    event_array = copy_array(v->data.array); \
+  } else { \
+    api_set_error( \
+        err, \
+        kErrorTypeValidation, \
+        "'event' must be an array or a string."); \
+    goto goto_name; \
+  }
+
+#define GET_ONE_EVENT(event_nr, event_str, goto_name) \
+  char_u *next_ev; \
+  event_T event_nr = event_name2nr((char_u *)event_str.data.string.data, &next_ev); \
+  if (event_nr >= NUM_EVENTS) { \
+    api_set_error(err, kErrorTypeValidation, "unexpected event"); \
+    goto goto_name; \
+  }
+
+
+/// Define an autocmd.
+/// @param opts Dictionary
+///          Required keys:
+///              event: string | ArrayOf(string)
+void nvim_autocmd_define(
+    DictionaryOf(LuaRef) opts,
+    Error *err)
+  FUNC_API_SINCE(7)
+{
+  char_u *rhs;
+  char_u *pattern = NULL;
+
+  bool is_once = false;
+  bool is_nested = false;
+  int au_group = AUGROUP_ALL;
+
+  LuaCallable command_callback = nlua_init_callable();
+  Array event_array = ARRAY_DICT_INIT;
+
+  bool found_command = false;
+
+  FOREACH_KEY_VALUE(opts, k, v, {
+    if (strequal("command", k.data)) {
```

Comment:
Hmm, commands and functions are really not the same thing, even within the same language. I think we should rather have `command="this is an ex command"` and `callback=luaref/funcref` separate exclusive keys. then if you pass `callback="MyFunction"` in vimscript it will get coerced to a funcref as expected.